### ISSUE TITLE
[HTTP] pipeline's status code vs wsi

### ISSUE THREAD

**Issue by ecionis** [warmcat/libwebsockets#1346]
Hi,

I want get http status code (via lws_http_client_http_response) with corresponding wsi and userdata. But, wsi is master.

Example callback (2 pipeline http reqs):
ctx: wsi_1 e: LWS_CALLBACK_ESTABLISHED_CLIENT_HTTP, lws_http_client_http_response:200
ctx: wsi_1 e: LWS_CALLBACK_RECEIVE_CLIENT_HTTP
ctx: wsi_1 e: LWS_CALLBACK_RECEIVE_CLIENT_HTTP_READ
ctx: wsi_1 e: LWS_CALLBACK_COMPLETED_CLIENT_HTTP
...
ctx: **wsi_1** e: LWS_CALLBACK_ESTABLISHED_CLIENT_HTTP, lws_http_client_http_response:200 (here is invalid ctx and user data)
ctx: **wsi_1** e: LWS_CALLBACK_RECEIVE_CLIENT_HTTP (here is invalid ctx and user data or not)
ctx: wsi_2 e: LWS_CALLBACK_RECEIVE_CLIENT_HTTP_READ
ctx: wsi_2 e: LWS_CALLBACK_COMPLETED_CLIENT_HTTP

I dont know how to fix it correctly.

For my question:
I add "wsi_eff=lws_client_wsi_effective(wsi)" for call LWS_CALLBACK_ESTABLISHED_CLIENT_HTTP (for valid ctx and userdata)
I add "_lws_client_wsi_master" to lws_http_client_http_response() for valid extract status code from master wsi.

**Comment by lws-team** [comment]
I kind of understand your general topic... but "invalid ctx"?  That can't be the lws_context, then what does it mean?

Can this problem be seen on the http-client multi example?

https://libwebsockets.org/git/libwebsockets/tree/minimal-examples/http-client/minimal-http-client-multi

**Comment by ecionis** [comment]
Oh, sorry
Invalid ctx, i mean. Ctx is wsi (with userdata/user_space), and when LWS_CALLBACK_ESTABLISHED_CLIENT_HTTP have been called for second request, wsi equal first request. And i dont have and link to second request to store his status code.

**Comment by lws-team** [comment]
Well... you can rename it how you like in your mind, but for any kind of communication I won't be able to understand you...

`lws_client_wsi_effective()` is not public IIRC.  It's used internally to find out who the original wsi who actually made the connection to the server is pipelining for now, after he completed his own transaction.  The idea is lws hides everything about whether http/1.1 pipelining connection queueing / coaslecing happened, or h2 connection binding happened, or they all connected individually.  The client wsis experience the same set of callbacks on their wsi either way.

If the callback feels something different, for me to understand the problem it will be very helpful if you can tell me how to see the problem on the example I linked above.  Because on that example you can control the connection binding mode precisely, and fetch 8 x simultaneous http things.  So if there's a problem like lws calls the master wsi for some callback accidentally, you should see it there with the pipelining switches enabled.

**Comment by ecionis** [comment]
I created 2 wsi with my user data  (class httpsocket).

example of process:

    int callback_http(lws *wsi, lws_callback_reasons reason, void *user, void *in, size_t len)
    {
        void *udata = lws_wsi_user(wsi);
        if(!udata || udata != user) return 0;        
        httpsocket* ctx = (httpsocket*)udata;
        switch(reason) 
        {
            case LWS_CALLBACK_ESTABLISHED_CLIENT_HTTP:
                ctx->last_http_status_code = (ui32)lws_http_client_http_response(wsi); //here is wrong user data for second request (because wsi=first request)
                break;

**Comment by lws-team** [comment]
Ok... I'll check it in the morning.

**Comment by lws-team** [comment]
I pushed this on master...

```
diff --git a/lib/roles/http/client/client.c b/lib/roles/http/client/client.c
index bb721ec7..6ca29d7f 100644
--- a/lib/roles/http/client/client.c
+++ b/lib/roles/http/client/client.c
@@ -914,9 +914,9 @@ lws_client_interpret_server_handshake(struct lws *wsi)
                 * we seem to be good to go, give client last chance to check
                 * headers and OK it
                 */
-               if (wsi->protocol->callback(wsi,
+               if (wsi->protocol->callback(w,
                                LWS_CALLBACK_CLIENT_FILTER_PRE_ESTABLISH,
-                                           wsi->user_space, NULL, 0)) {
+                                           w->user_space, NULL, 0)) {
 
                        cce = "HS: disallowed by client filter";
                        goto bail2;
@@ -928,9 +928,9 @@ lws_client_interpret_server_handshake(struct lws *wsi)
                wsi->rxflow_change_to = LWS_RXFLOW_ALLOW;
 
                /* call him back to inform him he is up */
-               if (wsi->protocol->callback(wsi,
+               if (wsi->protocol->callback(w,
                                            LWS_CALLBACK_ESTABLISHED_CLIENT_HTTP,
-                                           wsi->user_space, NULL, 0)) {
+                                           w->user_space, NULL, 0)) {
                        cce = "HS: disallowed at ESTABLISHED";
                        goto bail3;
                }
@@ -968,9 +968,9 @@ bail2:
                n = 0;
                if (cce)
                        n = (int)strlen(cce);
-               wsi->protocol->callback(wsi,
+               wsi->protocol->callback(w,
                                LWS_CALLBACK_CLIENT_CONNECTION_ERROR,
-                               wsi->user_space, (void *)cce,
+                               w->user_space, (void *)cce,
                                (unsigned int)n);
        }
        wsi->already_did_cce = 1;
```

Does it help?  Are there any other cases?

**Comment by ecionis** [comment]
Hi,

1. Its partial solution.  Status code was stored at master wsi. I did


lws_http_client_http_response(struct lws *_wsi)
{
    struct lws *wsi = _lws_client_wsi_master(_wsi); //<-- here is extract master wsi
	if (!wsi->http.ah) return 0;
	return wsi->http.ah->http_response;
}

2. Is "wsi->protocol->callback()" correct? may be "w->protocol->callback()"

**Comment by lws-team** [comment]
Thanks... yes on 2) they should also use the effective wsi protocol.

On 1) you're also right but http/1.1 pipelining is only one possible situation... with that only the original guy who connected gets the ah, and the same ah is reset and reused for each client... it ends up with the situation you found currently that the http response (which goes into the ah) can only be found by the ah connected to the "master".

But in either http/1.0 or h2, each client connection gets its own ah and they run simultaneously, then the ah belongs to the individual client connection and the http response is found in there.

I added logging the response code to the multi client minimal example, and confirmed this works with all the different ways

```
LWS_VISIBLE LWS_EXTERN unsigned int
lws_http_client_http_response(struct lws *_wsi)
{
	struct lws *wsi;
	unsigned int resp;

	if (_wsi->http.ah && _wsi->http.ah->http_response)
		return _wsi->http.ah->http_response;

	lws_vhost_lock(_wsi->vhost);
	wsi = _lws_client_wsi_master(_wsi);
	resp = wsi->http.ah->http_response;
	lws_vhost_unlock(_wsi->vhost);

	return resp;
}
```

I updated the patch on master to have that and with w->protocol fix and added the patch on v3.0-stable.
